fs/grub/disk_io.c

```diff
@@ -125,36 +125,30 @@ int filepos;
 int filemax;
 
 //Cromwell hooks
-int
-rawread (int drive, int sector, int byte_offset, int byte_len, char *buf)
+int rawread(int drive, int sector, int byte_offset, int byte_len, char *buf)
 {
-	sector+=byte_offset/512;
-	byte_offset%=512;
-	
-	while(byte_len) {
-		int nThisTime=512;
-	       	if(byte_len<512) nThisTime=byte_len;
-		if(byte_offset) {
-			u8 ba[512];
-			if(BootIdeReadSector(drive, buf, sector, 0, 512)) {
-				bprintf("Unable to get first sector\n");
-                             	return 1;
-			}
-			memcpy(buf, &ba[byte_offset], nThisTime-byte_offset);
-                      	buf+=nThisTime-byte_offset;
-                   	byte_len-=nThisTime-byte_offset;
-                      	byte_offset=0;
-             	} else {
-                     	if(BootIdeReadSector(drive, buf, sector, 0, nThisTime)) {
-                      		bprintf("Unable to get first sector\n");
-                             	return 1;
-                     	}
-                   	buf+=nThisTime;
-                   	byte_len-=nThisTime;
-              	}
+	// Advance sector counter by offset to get sub-sector-size offset
+	sector += byte_offset/SECTOR_SIZE;
+	byte_offset %= SECTOR_SIZE;
+
+	while (byte_len > 0) {
+		// Note: BootIdeReadSector will only read a single sector
+		int bytes_to_copy = MIN(byte_len, SECTOR_SIZE-byte_offset);
+		int status = BootIdeReadSector(drive, buf, sector, byte_offset, bytes_to_copy);
+		if (status != 0) {
+			bprintf("Unable to read sector %d of IDE drive %d (%d bytes)\n", sector, drive, bytes_to_copy);
+			return 1;
```

Comment:
This should ideally set an error message

See https://stuff.mit.edu/afs/sipb/project/hurddev/tb/grub-0.5.92/stage2/disk_io.c line 154:

```c
		    errnum = ERR_READ;
```

So this functions should set `errnum` in case of error, then `return 1`.

Reply:
Fixed.